In [480]:
import pandas as pd
import random as rd
import numpy as np

In [481]:
df = pd.read_csv("example.csv")

In [482]:
df

Unnamed: 0  Safety    yacEPA  Drive  FieldGoalDistance    airEPA  \
0          244485       0  1.240299     18                NaN -0.588101   
1          115340       0       NaN     22                NaN       NaN   
2           68357       0       NaN      8                NaN       NaN   
3          368377       0  1.075660     24                NaN  0.402080   
4          384684       0       NaN     11                NaN       NaN   
...           ...     ...       ...    ...                ...       ...   
40764       79438       0       NaN      3                NaN       NaN   
40765      398233       0       NaN      9                NaN       NaN   
40766      396466       0       NaN      8                NaN       NaN   
40767      333861       0  0.025040     14                NaN  0.280049   
40768      247608       0 -2.441755      6                NaN  1.692283   

       Field_Goal_Prob posteam  Penalty.Yards FieldGoalResult  ...  \
0             0.309404      TB              0             NaN  ...   
1             0.198142     OAK              0             NaN  ...   
2             0.000000     NYJ              0             NaN  ...   
3             0.258227     CLE              0             NaN  ...   
4             0.199771     DEN              0             NaN  ...   
...                ...     ...            ...             ...  ...   
40764         0.381506      NO              0             NaN  ...   
40765         0.244675     DEN              0             NaN  ...   
40766         0.147057     ATL              0             NaN  ...   
40767         0.373774     IND              0             NaN  ...   
40768         0.210732     CIN              0             NaN  ...   

       Accepted.Penalty qtr  AwayTimeouts_Remaining_Post  RunGap  Reception  \
0                     0   3                            3     NaN          1   
1                     0   4                            3     NaN          0   
2                     0   2                            1     NaN          0   
3                     0   4                            0     NaN          1   
4                     0   2                            2     end          0   
...                 ...  ..                          ...     ...        ...   
40764                 0   1                            3   guard          0   
40765                 0   2                            3   guard          0   
40766                 0   2                            1   guard          0   
40767                 0   3                            3     NaN          1   
40768                 0   1                            3     NaN          0   

       TimeUnder Season  Opp_Safety_Prob No_Score_Prob Challenge.Replay  
0              1   2014         0.000527      0.024134                0  
1              7   2011         0.001883      0.246929                0  
2              1   2010         0.000000      0.000000                0  
3              9   2017         0.000845      0.133814                0  
4             10   2017         0.003453      0.170787                0  
...          ...    ...              ...           ...              ...  
40764          3   2010         0.000021      0.005075                0  
40765          3   2017         0.000134      0.287528                0  
40766          1   2017         0.001132      0.612357                0  
40767          7   2016         0.000042      0.002641                0  
40768          2   2014         0.006138      0.040037                0  

[40769 rows x 32 columns]

In [483]:
df.isnull().sum()

Unnamed: 0                         0
Safety                             0
yacEPA                         24902
Drive                              0
FieldGoalDistance              39867
airEPA                         24894
Field_Goal_Prob                   21
posteam                         2550
Penalty.Yards                      0
FieldGoalResult                39861
DefTeamScore                    2750
Tackler2                       35826
Away_WP_pre                     2546
Timeout_Indicator                  0
ydsnet                             0
TimeSecs                          27
Passer                         24073
Touchdown                          0
TwoPointConv                   40716
PassOutcome                    24018
PlayTimeDiff                      43
RushAttempt                        0
Accepted.Penalty                   0
qtr                                0
AwayTimeouts_Remaining_Post        0
RunGap                         32008
Reception                          0
T

In [484]:
df.shape

(40769, 32)

In [485]:
df.columns

Index(['Unnamed: 0', 'Safety', 'yacEPA', 'Drive', 'FieldGoalDistance',
       'airEPA', 'Field_Goal_Prob', 'posteam', 'Penalty.Yards',
       'FieldGoalResult', 'DefTeamScore', 'Tackler2', 'Away_WP_pre',
       'Timeout_Indicator', 'ydsnet', 'TimeSecs', 'Passer', 'Touchdown',
       'TwoPointConv', 'PassOutcome', 'PlayTimeDiff', 'RushAttempt',
       'Accepted.Penalty', 'qtr', 'AwayTimeouts_Remaining_Post', 'RunGap',
       'Reception', 'TimeUnder', 'Season', 'Opp_Safety_Prob', 'No_Score_Prob',
       'Challenge.Replay'],
      dtype='object')

In [486]:
df['No_Score_Prob'] = df['No_Score_Prob'].fillna(value=df['No_Score_Prob'].median())
# df['No_Score_Prob']
df.isnull().sum()


Unnamed: 0                         0
Safety                             0
yacEPA                         24902
Drive                              0
FieldGoalDistance              39867
airEPA                         24894
Field_Goal_Prob                   21
posteam                         2550
Penalty.Yards                      0
FieldGoalResult                39861
DefTeamScore                    2750
Tackler2                       35826
Away_WP_pre                     2546
Timeout_Indicator                  0
ydsnet                             0
TimeSecs                          27
Passer                         24073
Touchdown                          0
TwoPointConv                   40716
PassOutcome                    24018
PlayTimeDiff                      43
RushAttempt                        0
Accepted.Penalty                   0
qtr                                0
AwayTimeouts_Remaining_Post        0
RunGap                         32008
Reception                          0
T

In [487]:
df['Opp_Safety_Prob'] = df['Opp_Safety_Prob'].fillna(value=df['Opp_Safety_Prob'].median())
df.isnull().sum()


Unnamed: 0                         0
Safety                             0
yacEPA                         24902
Drive                              0
FieldGoalDistance              39867
airEPA                         24894
Field_Goal_Prob                   21
posteam                         2550
Penalty.Yards                      0
FieldGoalResult                39861
DefTeamScore                    2750
Tackler2                       35826
Away_WP_pre                     2546
Timeout_Indicator                  0
ydsnet                             0
TimeSecs                          27
Passer                         24073
Touchdown                          0
TwoPointConv                   40716
PassOutcome                    24018
PlayTimeDiff                      43
RushAttempt                        0
Accepted.Penalty                   0
qtr                                0
AwayTimeouts_Remaining_Post        0
RunGap                         32008
Reception                          0
T

In [488]:
df['PlayTimeDiff']

0        12.0
1        44.0
2         0.0
3         8.0
4         0.0
         ... 
40764    31.0
40765    20.0
40766    27.0
40767     7.0
40768     5.0
Name: PlayTimeDiff, Length: 40769, dtype: float64

In [489]:
print(df['PlayTimeDiff'].mean())
print(df['PlayTimeDiff'].median())
print(df['PlayTimeDiff'].min())
print(df['PlayTimeDiff'].max())


20.77422285517851
18.0
0.0
927.0


In [490]:
df['PlayTimeDiff']= df['PlayTimeDiff'].fillna(value=df['PlayTimeDiff'].median())

In [491]:
df.isnull().sum()

Unnamed: 0                         0
Safety                             0
yacEPA                         24902
Drive                              0
FieldGoalDistance              39867
airEPA                         24894
Field_Goal_Prob                   21
posteam                         2550
Penalty.Yards                      0
FieldGoalResult                39861
DefTeamScore                    2750
Tackler2                       35826
Away_WP_pre                     2546
Timeout_Indicator                  0
ydsnet                             0
TimeSecs                          27
Passer                         24073
Touchdown                          0
TwoPointConv                   40716
PassOutcome                    24018
PlayTimeDiff                       0
RushAttempt                        0
Accepted.Penalty                   0
qtr                                0
AwayTimeouts_Remaining_Post        0
RunGap                         32008
Reception                          0
T

In [492]:
df['PassOutcome']

0               Complete
1                    NaN
2                    NaN
3               Complete
4                    NaN
              ...       
40764                NaN
40765                NaN
40766                NaN
40767           Complete
40768    Incomplete Pass
Name: PassOutcome, Length: 40769, dtype: object

In [493]:
print(df['PassOutcome'].value_counts())
print(df['PassOutcome'].isnull().sum())


PassOutcome
Complete           10122
Incomplete Pass     6629
Name: count, dtype: int64
24018


## PassOutcome ustunidagi valuelarni hamda NAn qiymatlarni sanab oldim

In [494]:
df['PassOutcome'] = df['PassOutcome'].fillna(value='-1')


## ikki hil javob bor Complete , Incomplete pass va NaN qiymatlar ikkisiining yigindisidan ham oshib ketadi shuning uchun men NAN qiymatlar orniga ikkisidan birini qoyishuim juda katta xatolik boladi deb oyladim va malumot berilmagani uchun NAN qiymatlar orniga missin indiktor qilib ushbu ustunda umuman uchramydigan -1 ni qoyib ketdim

In [495]:
print(df['PassOutcome'].value_counts())
print(df['PassOutcome'].isnull().sum())

PassOutcome
-1                 24018
Complete           10122
Incomplete Pass     6629
Name: count, dtype: int64
0


## va buyerda qilgan ishimni tekshirdim

In [496]:
df.isnull().sum()

Unnamed: 0                         0
Safety                             0
yacEPA                         24902
Drive                              0
FieldGoalDistance              39867
airEPA                         24894
Field_Goal_Prob                   21
posteam                         2550
Penalty.Yards                      0
FieldGoalResult                39861
DefTeamScore                    2750
Tackler2                       35826
Away_WP_pre                     2546
Timeout_Indicator                  0
ydsnet                             0
TimeSecs                          27
Passer                         24073
Touchdown                          0
TwoPointConv                   40716
PassOutcome                        0
PlayTimeDiff                       0
RushAttempt                        0
Accepted.Penalty                   0
qtr                                0
AwayTimeouts_Remaining_Post        0
RunGap                         32008
Reception                          0
T

## tekshirish uchun

In [497]:
print(df['TwoPointConv'].value_counts())
print(df['TwoPointConv'].isnull().sum())

TwoPointConv
Failure    28
Success    25
Name: count, dtype: int64
40716


## bu yerda ham shu holat kuzatilmoqda

In [498]:
df['TwoPointConv'] = df['TwoPointConv'].fillna(value='-1')

In [499]:
print(df['TwoPointConv'].value_counts())
print(df['TwoPointConv'].isnull().sum())
print(df.isnull().sum())

TwoPointConv
-1         40716
Failure       28
Success       25
Name: count, dtype: int64
0
Unnamed: 0                         0
Safety                             0
yacEPA                         24902
Drive                              0
FieldGoalDistance              39867
airEPA                         24894
Field_Goal_Prob                   21
posteam                         2550
Penalty.Yards                      0
FieldGoalResult                39861
DefTeamScore                    2750
Tackler2                       35826
Away_WP_pre                     2546
Timeout_Indicator                  0
ydsnet                             0
TimeSecs                          27
Passer                         24073
Touchdown                          0
TwoPointConv                       0
PassOutcome                        0
PlayTimeDiff                       0
RushAttempt                        0
Accepted.Penalty                   0
qtr                                0
AwayTimeouts_Remaini

## menda yana tepadagi holat yuz berdi va men yana shunday ishni amalga oshirdim

In [500]:
print(df['Passer'].value_counts())
print(df['Passer'].isnull().sum())

Passer
D.Brees         577
E.Manning       541
M.Stafford      534
T.Brady         523
P.Rivers        514
               ... 
C.Cook            1
M.Koenen          1
A.Brown           1
L.Fitzgerald      1
J.Callahan        1
Name: count, Length: 187, dtype: int64
24073


In [501]:
print(df['Passer'].fillna(method='ffill').value_counts())
# df.value_counts()

Passer
D.Brees         1417
E.Manning       1329
T.Brady         1296
P.Rivers        1290
M.Stafford      1265
                ... 
D.Robinson         1
M.Koenen           1
C.Cook             1
A.Brown            1
L.Fitzgerald       1
Name: count, Length: 187, dtype: int64


C:\Users\User\AppData\Local\Temp\ipykernel_14012\3125964819.py:1: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  print(df['Passer'].fillna(method='ffill').value_counts())


## men togrisi 40000 mingta ustundan 24000 tasi bosh bolgan ismlar qatorini qanday yol bilan toldirish togri ekanini bilmadim va buning tollari sifattida bir nechtasini keltirib otaman va oxirgisi bu men optimal deb topgan yol boladi va data baza shunday ozgarish bilan davom etadi tepadagi yol ozidan oldingi va keyingi ismlar bilan toldirish usuli

In [502]:
names = df['Passer'].unique()
names.tolist()
print(df['Passer'].fillna(lambda x: np.random.choice(names) if pd.isna(x) else x))
df['Passer'].value_counts()

0                                        M.Glennon
1        <function <lambda> at 0x0000017E1C3936A0>
2        <function <lambda> at 0x0000017E1C3936A0>
3                                          D.Kizer
4        <function <lambda> at 0x0000017E1C3936A0>
                           ...                    
40764    <function <lambda> at 0x0000017E1C3936A0>
40765    <function <lambda> at 0x0000017E1C3936A0>
40766    <function <lambda> at 0x0000017E1C3936A0>
40767                                       A.Luck
40768                                     A.Dalton
Name: Passer, Length: 40769, dtype: object


Passer
D.Brees         577
E.Manning       541
M.Stafford      534
T.Brady         523
P.Rivers        514
               ... 
C.Cook            1
M.Koenen          1
A.Brown           1
L.Fitzgerald      1
J.Callahan        1
Name: count, Length: 187, dtype: int64

## 

## uchraydigan ismlarni listga yigib olib NAN qiymatlarni orniga random orqali tanlatish usuli

In [503]:
df['Passer'] = df['Passer'].fillna(value='Not Given')
df['Passer'].value_counts()

Passer
Not Given       24073
D.Brees           577
E.Manning         541
M.Stafford        534
T.Brady           523
                ...  
C.Cook              1
M.Koenen            1
A.Brown             1
L.Fitzgerald        1
J.Callahan          1
Name: count, Length: 188, dtype: int64

## men shu yolni tanladim sababi buncha malumotlarni ozimizning random yoki boshqa yo bila toldirishimiz katta xatoga olib keladi deb oylayman garchi bu ismlar bolsa ham men ularni shunday yol bilan toldirish yaxshiroq deb oylayman

In [504]:
print(df.isnull().sum())


Unnamed: 0                         0
Safety                             0
yacEPA                         24902
Drive                              0
FieldGoalDistance              39867
airEPA                         24894
Field_Goal_Prob                   21
posteam                         2550
Penalty.Yards                      0
FieldGoalResult                39861
DefTeamScore                    2750
Tackler2                       35826
Away_WP_pre                     2546
Timeout_Indicator                  0
ydsnet                             0
TimeSecs                          27
Passer                             0
Touchdown                          0
TwoPointConv                       0
PassOutcome                        0
PlayTimeDiff                       0
RushAttempt                        0
Accepted.Penalty                   0
qtr                                0
AwayTimeouts_Remaining_Post        0
RunGap                         32008
Reception                          0
T

## ozgarishni tekshirish uchun

In [505]:
print(df['Away_WP_pre'].isnull().sum())
print(df['Away_WP_pre'].value_counts())
print(df['Away_WP_pre'].mean())
print(df['Away_WP_pre'].median())
print(df['Away_WP_pre'])
print(df['Away_WP_pre'].std())


2546
Away_WP_pre
0.514325    141
0.485675    126
0.000000    116
1.000000    103
0.499993     45
           ... 
0.163670      1
0.139549      1
0.467202      1
0.773475      1
0.000254      1
Name: count, Length: 37264, dtype: int64
0.46869555198829244
0.471362239854573
0        0.774353
1        0.943964
2        0.634693
3        0.064005
4        0.071526
           ...   
40764    0.163670
40765    0.139549
40766    0.467202
40767    0.773475
40768    0.613732
Name: Away_WP_pre, Length: 40769, dtype: float64
0.2857741835257317


## men bu ustunni ozini va mediana , avg va valuelari sonini kordim tahlil uchun

In [506]:
df['Away_WP_pre'] = df['Away_WP_pre'].fillna(value=df['Away_WP_pre'].median())
print(df['Away_WP_pre'].value_counts())
print(df.isnull().sum())
print(df['Away_WP_pre'].std())


Away_WP_pre
0.471362    2547
0.514325     141
0.485675     126
0.000000     116
1.000000     103
            ... 
0.163670       1
0.139549       1
0.467202       1
0.773475       1
0.128579       1
Name: count, Length: 37264, dtype: int64
Unnamed: 0                         0
Safety                             0
yacEPA                         24902
Drive                              0
FieldGoalDistance              39867
airEPA                         24894
Field_Goal_Prob                   21
posteam                         2550
Penalty.Yards                      0
FieldGoalResult                39861
DefTeamScore                    2750
Tackler2                       35826
Away_WP_pre                        0
Timeout_Indicator                  0
ydsnet                             0
TimeSecs                          27
Passer                             0
Touchdown                          0
TwoPointConv                       0
PassOutcome                        0
PlayTimeDiff        

## 0.27670690295437705 bu son men ortacha qiymatni NAN lar uchunqoyganimda paydo bolgan std tepada esa mediana qoyganimda ozgarishni korishingiz mumkin 0.2857741835257317 bu esa ozgarishdan oldingi std men qilgan ozgarish unga oz miqdorda yaqin bolgani uchun men medianani tanladim

In [507]:
df.isnull().sum()

Unnamed: 0                         0
Safety                             0
yacEPA                         24902
Drive                              0
FieldGoalDistance              39867
airEPA                         24894
Field_Goal_Prob                   21
posteam                         2550
Penalty.Yards                      0
FieldGoalResult                39861
DefTeamScore                    2750
Tackler2                       35826
Away_WP_pre                        0
Timeout_Indicator                  0
ydsnet                             0
TimeSecs                          27
Passer                             0
Touchdown                          0
TwoPointConv                       0
PassOutcome                        0
PlayTimeDiff                       0
RushAttempt                        0
Accepted.Penalty                   0
qtr                                0
AwayTimeouts_Remaining_Post        0
RunGap                         32008
Reception                          0
T

In [508]:
print(df['TimeSecs'].value_counts())
print(df['TimeSecs'].mean())
# df['TimeSecs'].isnull().sum()
df['TimeSecs'].std()

TimeSecs
 1800.0    508
 900.0     410
 2700.0    408
 3600.0    360
 120.0     360
          ... 
-91.0        1
 3598.0      1
-407.0       1
-543.0       1
-803.0       1
Name: count, Length: 3825, dtype: int64
1699.549874822051


np.float64(1063.620262887654)

## tahlil uchun


In [509]:
print(df['TimeSecs'].fillna(value=df['TimeSecs'].mean()))
print(df['TimeSecs'].std())
df['TimeSecs'] = df['TimeSecs'].fillna(value=df['TimeSecs'].mean())
print(df['TimeSecs'].value_counts())
print(df['TimeSecs'].std())
df['TimeSecs'].isnull().sum()

0         939.0
1         407.0
2        1823.0
3         528.0
4        2355.0
          ...  
40764    2868.0
40765    1962.0
40766    1856.0
40767    1289.0
40768    2776.0
Name: TimeSecs, Length: 40769, dtype: float64
1063.620262887654
TimeSecs
 1800.0    508
 900.0     410
 2700.0    408
 120.0     360
 3600.0    360
          ... 
-155.0       1
-308.0       1
-294.0       1
-169.0       1
-41.0        1
Name: count, Length: 3826, dtype: int64
1063.2679951347561


np.int64(0)

In [510]:
df.isnull().sum()

Unnamed: 0                         0
Safety                             0
yacEPA                         24902
Drive                              0
FieldGoalDistance              39867
airEPA                         24894
Field_Goal_Prob                   21
posteam                         2550
Penalty.Yards                      0
FieldGoalResult                39861
DefTeamScore                    2750
Tackler2                       35826
Away_WP_pre                        0
Timeout_Indicator                  0
ydsnet                             0
TimeSecs                           0
Passer                             0
Touchdown                          0
TwoPointConv                       0
PassOutcome                        0
PlayTimeDiff                       0
RushAttempt                        0
Accepted.Penalty                   0
qtr                                0
AwayTimeouts_Remaining_Post        0
RunGap                         32008
Reception                          0
T

In [511]:
df = df.drop(columns='Tackler2')
df.isnull().sum()


Unnamed: 0                         0
Safety                             0
yacEPA                         24902
Drive                              0
FieldGoalDistance              39867
airEPA                         24894
Field_Goal_Prob                   21
posteam                         2550
Penalty.Yards                      0
FieldGoalResult                39861
DefTeamScore                    2750
Away_WP_pre                        0
Timeout_Indicator                  0
ydsnet                             0
TimeSecs                           0
Passer                             0
Touchdown                          0
TwoPointConv                       0
PassOutcome                        0
PlayTimeDiff                       0
RushAttempt                        0
Accepted.Penalty                   0
qtr                                0
AwayTimeouts_Remaining_Post        0
RunGap                         32008
Reception                          0
TimeUnder                          0
S

## Tackler2 columinida 87 foizdan koproq malumotlar yoqligi uchun uni xatoliklar keltirib chiqarmasligi uchun ochirib tashladim

In [512]:
df = df.drop(columns='FieldGoalResult')
df.isnull().sum()

Unnamed: 0                         0
Safety                             0
yacEPA                         24902
Drive                              0
FieldGoalDistance              39867
airEPA                         24894
Field_Goal_Prob                   21
posteam                         2550
Penalty.Yards                      0
DefTeamScore                    2750
Away_WP_pre                        0
Timeout_Indicator                  0
ydsnet                             0
TimeSecs                           0
Passer                             0
Touchdown                          0
TwoPointConv                       0
PassOutcome                        0
PlayTimeDiff                       0
RushAttempt                        0
Accepted.Penalty                   0
qtr                                0
AwayTimeouts_Remaining_Post        0
RunGap                         32008
Reception                          0
TimeUnder                          0
Season                             0
O

## FieldGoalResult ustunida 97 foizdan koproq malumot yoq bolgani uchun uni ham yoqotdim

In [513]:
df = df.drop(columns='FieldGoalDistance')
df.isnull().sum()

Unnamed: 0                         0
Safety                             0
yacEPA                         24902
Drive                              0
airEPA                         24894
Field_Goal_Prob                   21
posteam                         2550
Penalty.Yards                      0
DefTeamScore                    2750
Away_WP_pre                        0
Timeout_Indicator                  0
ydsnet                             0
TimeSecs                           0
Passer                             0
Touchdown                          0
TwoPointConv                       0
PassOutcome                        0
PlayTimeDiff                       0
RushAttempt                        0
Accepted.Penalty                   0
qtr                                0
AwayTimeouts_Remaining_Post        0
RunGap                         32008
Reception                          0
TimeUnder                          0
Season                             0
Opp_Safety_Prob                    0
N

## FieldGoalDistance ustunida ham 97 foizdaan koproq malumotlar yoq va u ham ochirildi

In [514]:
df = df.drop(columns='RunGap')
df.isnull().sum()

Unnamed: 0                         0
Safety                             0
yacEPA                         24902
Drive                              0
airEPA                         24894
Field_Goal_Prob                   21
posteam                         2550
Penalty.Yards                      0
DefTeamScore                    2750
Away_WP_pre                        0
Timeout_Indicator                  0
ydsnet                             0
TimeSecs                           0
Passer                             0
Touchdown                          0
TwoPointConv                       0
PassOutcome                        0
PlayTimeDiff                       0
RushAttempt                        0
Accepted.Penalty                   0
qtr                                0
AwayTimeouts_Remaining_Post        0
Reception                          0
TimeUnder                          0
Season                             0
Opp_Safety_Prob                    0
No_Score_Prob                      0
C

## RunGap ustunida ham 78 foizdan koproq malumotlar yoqligi tufayli ochirildi

In [515]:
print(df['DefTeamScore'].isnull().sum())
print(df['DefTeamScore'].value_counts())
print(df['DefTeamScore'].mean())
# print(df['DefTeamScore'].min())
# print(df['DefTeamScore'].max())
print(df['DefTeamScore'].std())
print(df['DefTeamScore'].median())


2750
DefTeamScore
0.0     7779
7.0     4508
3.0     3545
10.0    3287
14.0    2159
17.0    2113
13.0    1851
6.0     1650
20.0    1444
24.0    1063
21.0     957
16.0     879
23.0     779
9.0      723
27.0     697
19.0     479
31.0     460
28.0     395
26.0     329
30.0     324
12.0     312
34.0     281
22.0     215
33.0     174
15.0     170
35.0     167
38.0     166
29.0     134
37.0     105
41.0      91
18.0      80
25.0      78
36.0      76
42.0      73
8.0       63
32.0      54
2.0       51
5.0       42
40.0      41
45.0      40
11.0      27
39.0      25
48.0      24
44.0      23
43.0      20
49.0      14
52.0       8
46.0       7
47.0       6
54.0       5
1.0        4
55.0       4
50.0       4
51.0       3
4.0        3
58.0       3
59.0       2
53.0       1
56.0       1
61.0       1
Name: count, dtype: int64
11.366316841579211
9.90134048283482
10.0


In [516]:
df['DefTeamScore'] = df['DefTeamScore'].fillna(value=df['DefTeamScore'].median())
print(df['DefTeamScore'].isnull().sum())

df['DefTeamScore'].std()

0


np.float64(9.567702623348124)

## ortacha qiymat qoyganimda 9.561563753208729 shu natija va mediana qoyganimda tepadagi natija chiqdi va std ikkilasini qoyib tekshirganimda mediananiki yaqin bolgani uchun uni qoydim

In [517]:
df.isna().sum()

Unnamed: 0                         0
Safety                             0
yacEPA                         24902
Drive                              0
airEPA                         24894
Field_Goal_Prob                   21
posteam                         2550
Penalty.Yards                      0
DefTeamScore                       0
Away_WP_pre                        0
Timeout_Indicator                  0
ydsnet                             0
TimeSecs                           0
Passer                             0
Touchdown                          0
TwoPointConv                       0
PassOutcome                        0
PlayTimeDiff                       0
RushAttempt                        0
Accepted.Penalty                   0
qtr                                0
AwayTimeouts_Remaining_Post        0
Reception                          0
TimeUnder                          0
Season                             0
Opp_Safety_Prob                    0
No_Score_Prob                      0
C

In [518]:
df.columns

Index(['Unnamed: 0', 'Safety', 'yacEPA', 'Drive', 'airEPA', 'Field_Goal_Prob',
       'posteam', 'Penalty.Yards', 'DefTeamScore', 'Away_WP_pre',
       'Timeout_Indicator', 'ydsnet', 'TimeSecs', 'Passer', 'Touchdown',
       'TwoPointConv', 'PassOutcome', 'PlayTimeDiff', 'RushAttempt',
       'Accepted.Penalty', 'qtr', 'AwayTimeouts_Remaining_Post', 'Reception',
       'TimeUnder', 'Season', 'Opp_Safety_Prob', 'No_Score_Prob',
       'Challenge.Replay'],
      dtype='object')

In [519]:
df.shape

(40769, 28)

In [520]:
print(df['posteam'].isna().sum())
print(df['posteam'].value_counts())

2550
posteam
DEN    1284
PHI    1244
DET    1243
NO     1233
BAL    1218
IND    1218
NE     1216
SEA    1211
KC     1210
HOU    1209
ARI    1208
SF     1207
WAS    1201
GB     1198
DAL    1192
CIN    1189
CHI    1188
BUF    1186
NYJ    1183
MIN    1182
NYG    1181
OAK    1180
ATL    1167
MIA    1160
TB     1149
PIT    1147
TEN    1138
CAR    1128
CLE    1117
SD     1086
STL     968
JAC     936
JAX     272
LA      255
LAC     115
Name: count, dtype: int64


In [521]:
df['posteam'] = df['posteam'].fillna(value='NOT GIVEN')

In [522]:
print(df['posteam'].isna().sum())
print(df['posteam'].value_counts())

0
posteam
NOT GIVEN    2550
DEN          1284
PHI          1244
DET          1243
NO           1233
IND          1218
BAL          1218
NE           1216
SEA          1211
KC           1210
HOU          1209
ARI          1208
SF           1207
WAS          1201
GB           1198
DAL          1192
CIN          1189
CHI          1188
BUF          1186
NYJ          1183
MIN          1182
NYG          1181
OAK          1180
ATL          1167
MIA          1160
TB           1149
PIT          1147
TEN          1138
CAR          1128
CLE          1117
SD           1086
STL           968
JAC           936
JAX           272
LA            255
LAC           115
Name: count, dtype: int64


## men tepadagi posteam columni ustida qilgan ishlarim shu boldiki yani men yoq bolgan 2550 ta malumot orniga NOT GIVEN deb berib qoydim, nimaga chunki buncha miqdordagi malumotni qaysidir yo bilan qolda toldirish xatoliklarga olib keladi chunki bu ustun bilan grafik qilinganda ular tahlil keskin buzib yuborishi mumkin, Axir bu 2550 ta malumoyt barcha malumotlarning deyarli 1/16 qismini tashkil qiladi 

In [523]:
df.isna().sum()

Unnamed: 0                         0
Safety                             0
yacEPA                         24902
Drive                              0
airEPA                         24894
Field_Goal_Prob                   21
posteam                            0
Penalty.Yards                      0
DefTeamScore                       0
Away_WP_pre                        0
Timeout_Indicator                  0
ydsnet                             0
TimeSecs                           0
Passer                             0
Touchdown                          0
TwoPointConv                       0
PassOutcome                        0
PlayTimeDiff                       0
RushAttempt                        0
Accepted.Penalty                   0
qtr                                0
AwayTimeouts_Remaining_Post        0
Reception                          0
TimeUnder                          0
Season                             0
Opp_Safety_Prob                    0
No_Score_Prob                      0
C

In [524]:
print(df['airEPA'].value_counts())
print(df['airEPA'].mean())
print(df['airEPA'].median())
print(df['airEPA'].std())


airEPA
 1.053000    17
 0.000000     8
 0.084494     6
 0.969958     5
-1.469790     5
             ..
 0.754171     1
 1.418653     1
-0.107088     1
 0.980268     1
-0.825221     1
Name: count, Length: 15788, dtype: int64
0.5223888106338055
0.2956190905026767
1.399719636710543


In [525]:
df['airEPA'] = df['airEPA'].fillna(value='-9999')
print(df['airEPA'].value_counts())


airEPA
 -9999       24894
 1.053          17
 0.0             8
 0.084494        6
-1.46979         5
             ...  
 0.754171        1
 1.418653        1
-0.107088        1
 0.980268        1
-0.103057        1
Name: count, Length: 15789, dtype: int64


## airEPA ustunida berilmagan malumotlar 70 va 80 foizdan oshmagani uchun uni ochirmadim va NAN qiymatlarni yoqotish uchun missing indikator sifatida ustunda uchramaydigan -9999 ni qoyib ketdim

In [526]:
print(df['yacEPA'].isnull().sum())
print(df['yacEPA'].value_counts())
print(df['yacEPA'].mean())
print(df['yacEPA'].median())
print(df['yacEPA'].std())

24902
yacEPA
 0.000000     2366
-7.000000       13
-0.864491        3
-14.000000       3
-0.323526        2
              ... 
 0.775660        1
-10.853876       1
 1.704423        1
-0.485422        1
 1.075660        1
Name: count, Length: 13480, dtype: int64
-0.39613653499874507
0.0
1.9986634244316326


In [527]:
df['yacEPA'] = df['yacEPA'].fillna(value='-9999')
df.isnull().sum()

Unnamed: 0                      0
Safety                          0
yacEPA                          0
Drive                           0
airEPA                          0
Field_Goal_Prob                21
posteam                         0
Penalty.Yards                   0
DefTeamScore                    0
Away_WP_pre                     0
Timeout_Indicator               0
ydsnet                          0
TimeSecs                        0
Passer                          0
Touchdown                       0
TwoPointConv                    0
PassOutcome                     0
PlayTimeDiff                    0
RushAttempt                     0
Accepted.Penalty                0
qtr                             0
AwayTimeouts_Remaining_Post     0
Reception                       0
TimeUnder                       0
Season                          0
Opp_Safety_Prob                 0
No_Score_Prob                   0
Challenge.Replay                0
dtype: int64

## men yacEPA ustuni boyicha ham shunday hulosaga keldim va unga ham missing indikator sifatida -9999 berib ketdim

In [528]:
print(df['Field_Goal_Prob'].isnull().sum())
print(df['Field_Goal_Prob'].value_counts())
print(df['Field_Goal_Prob'].mean())
print(df['Field_Goal_Prob'].median())
print(df['Field_Goal_Prob'].std())

21
Field_Goal_Prob
0.000000    4096
0.213700     544
0.233081     174
0.208219      23
0.208102       9
            ... 
0.432242       1
0.381506       1
0.244675       1
0.147057       1
0.272976       1
Name: count, Length: 33822, dtype: int64
0.24367961032922636
0.2306319714387671
0.15940004174754327


In [529]:
df['Field_Goal_Prob'] = df['Field_Goal_Prob'].fillna(value=df['Field_Goal_Prob'].median())


## men yoq bolgan 21 malumot orniga ustun medianasini qoyib ketdim

In [530]:
df.head(60)

Unnamed: 0  Safety    yacEPA  Drive    airEPA  Field_Goal_Prob    posteam  \
0       244485       0  1.240299     18 -0.588101         0.309404         TB   
1       115340       0     -9999     22     -9999         0.198142        OAK   
2        68357       0     -9999      8     -9999         0.000000        NYJ   
3       368377       0   1.07566     24   0.40208         0.258227        CLE   
4       384684       0     -9999     11     -9999         0.199771        DEN   
5       277250       0     -9999     20     -9999         0.000000  NOT GIVEN   
6       398793       0       0.0     20  0.081113         0.272976        SEA   
7       370285       0       0.0      9 -2.611042         0.395233        LAC   
8       364538       0  0.477721      1 -0.095133         0.317982        DAL   
9       229159       0     -9999     15     -9999         0.000000         NO   
10      357254       0     -9999     20     -9999         0.539731         NE   
11      246343       0     -9999     25     -9999         0.118200        OAK   
12      216491       0  1.454289      8 -0.657586         0.320544         NO   
13      342251       0     -9999     16     -9999         0.000000         GB   
14      161472       0     -9999     21     -9999         0.256153        WAS   
15      348211       0     -9999     13     -9999         0.000000  NOT GIVEN   
16      325513       0     -9999      2     -9999         0.121411        JAX   
17       48918       0     -9999      1     -9999         0.287326        STL   
18      244580       0     -9999      9     -9999         0.229492        NYJ   
19       37441       0     -9999     20     -9999         0.120585        MIA   
20      140128       0     -9999     10     -9999         0.070298        CAR   
21       19773       0  0.383471     25  0.123643         0.174577        ARI   
22      380092       0     -9999      7     -9999         0.000000  NOT GIVEN   
23       33174       0     -9999      7     -9999         0.278403        ARI   
24       47871       0     -9999     20     -9999         0.227382        CAR   
25      224824       0     -9999     16     -9999         0.263293         SD   
26       25942       0     -9999     19     -9999         0.104813        MIN   
27       87227       0 -2.110818     22  0.539738         0.361117        CIN   
28      388888       0     -9999      1     -9999         0.315067         GB   
29       28360       0     -9999     17     -9999         0.421281        DEN   
30      133668       0     -9999     20     -9999         0.187521         KC   
31       25896       0  1.296887     15  2.682991         0.109701        MIN   
32        7711       0     -9999      7     -9999         0.000000  NOT GIVEN   
33      116940       0  3.022639     20 -0.825221         0.080754        CHI   
34      273431       0     -9999      4     -9999         0.147104         KC   
35      278394       0     -9999      8     -9999         0.168599         SF   
36      273171       0   0.64166     19 -0.103057         0.241818        PHI   
37      351023       0 -3.265097      7  2.548888         0.429965        NYJ   
38      282554       0     -9999      5     -9999         0.000000        WAS   
39       81360       0     -9999      2     -9999         0.252942        BAL   
40      104418       0 -0.620298     14  0.137978         0.434540        BAL   
41      215319       0     -9999      9     -9999         0.382055         TB   
42      204981       0     -9999     31     -9999         0.330422        BAL   
43      338274       0  0.026307     11  1.354832         0.314212        MIN   
44      391308       0 -0.000954     21 -0.774933         0.320003        WAS   
45      223675       0     -9999     18     -9999         0.544393        ATL   
46       70649       0 -0.882931     21 -0.627575         0.312642        TEN   
47      205459       0  1.602694     21 -1.084424         0.307128        ARI   
48      186773       0 -7.421689   

## Va qarshingizda tayyor bolgan data set